### PR TITLE
Update minimal CMake version in favour of advances features it offers

### PR DESCRIPTION
**Pull request by oshadura**
Fixes as well https://github.com/root-project/root/issues/8280

(Thanks a lot to @amadio for pointing out the most reasonable for ROOT build system)

Update: sadly we can't move directly to 3.16 because Debian 10 has  only cmake 3.13 packaged https://pkgs.org/search/?q=cmake

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -11,7 +11,7 @@ if(MSVC)
   #       are now updated only when the object files providing the symbols are updated
   cmake_minimum_required(VERSION 3.16 FATAL_ERROR)
 else()
-  cmake_minimum_required(VERSION 3.9 FATAL_ERROR)
+  cmake_minimum_required(VERSION 3.12 FATAL_ERROR)
```
Comment (amadio) [inline CMakeLists.txt:14]: Wasn't this supposed to be 3.16? Also, I supposed in the same PR one should search and remove hacks for older versions (like the Python setup), as that will not be necessary anymore.

Comment (oshadura) [reply CMakeLists.txt]: @amadio sadly we can't move directly to 3.16 because Debian 10 has only cmake 3.13 packaged https://pkgs.org/search/?q=cmake

Comment (amadio) [reply CMakeLists.txt]: Ok, if you want to keep Debian 10 support, I guess 3.12 is the minimal version to be worth an update.

Comment (oshadura) [reply CMakeLists.txt]: I updated to 3.13 which is a bit closer to 3.16

**Comment by oshadura** [comment]
@amadio looks like it will be still impossible to remove Python hacks in https://github.com/root-project/root/blob/master/cmake/modules/SearchRootCoreDeps.cmake: `CMake 3.14 is the minimum because in 3.12 and 3.13, despite find_package(PythonX) being present, the Numpy-related variables are not set`

File: CMakeLists.txt (hunk 1 above)
Comment (Axel-Naumann) [reply CMakeLists.txt]: I'd find it reasonable to install CMake 3.16 on our Debian machines, and to expect packagers to have to do the same. After all it's not a runtime requirement.

Can people still build *against* ROOT with older CMake?

**Comment by amadio** [comment]
Yes, I was thinking about the move happening to 3.16. If it's only to 3.12 or 3.13, then the hacks need to stay. However, it's probably better to drop Debian 10 support and go to CMake 3.16. Debian 10 users can still have up to ROOT 6.24 if they don't want to install CMake by hand.

**Comment by oshadura** [comment]
About debian looks like users can get higher cmake version from testing repo: https://tracker.debian.org/pkg/cmake

File: CMakeLists.txt (hunk 1 above)
Comment (oshadura) [reply CMakeLists.txt]: I believe yes, but we will need to remove hacks https://github.com/root-project/root/issues/8280

Comment (oshadura) [reply CMakeLists.txt]: (I will test such combination with rootest.git)

**Comment by oshadura** [comment]
@phsft-bot build!

**Comment by Axel-Naumann** [comment]
@phsft-bot build just on ROOT-debian10-i386/cxx14

**Comment by bellenot** [comment]
So IIUC, Mac and Debian nodes should be updated with CMake 3.16

**Comment by oshadura** [comment]
So there should only one real failure mac11.0/cxx17. What is special in this setup? Does it have python3 available or only python2?

**Comment by oshadura** [comment]
@Axel-Naumann @bellenot  I think I know why `mac11.0/cxx17` is failing.
I see it is requested in list of  `all modules`  that `pyroot3` should be enabled, while Python3  is not available/installed on machine: 
```AXEL: all modules = alien;arrow;asimage;asserts;builtin_afterimage;builtin_cfitsio;builtin_davix;builtin_fftw3;builtin_freetype;builtin_ftgl;builtin_gl2ps;builtin_glew;builtin_gsl;builtin_lz4;builtin_lzma;builtin_nlohmannjson;builtin_openssl;builtin_openui5;builtin_pcre;builtin_tbb;builtin_unuran;builtin_vc;builtin_vdt;builtin_veccore;builtin_xrootd;builtin_xxhash;builtin_zlib;builtin_zstd;cefweb;clad;cocoa;cuda;cudnn;daos;dataframe;davix;dcache;dev;distcc;fcgi;fftw3;fitsio;fortran;gdml;gfal;gsl_shared;gviz;http;imt;libcxx;mathmore;minuit2;mlp;monalisa;mpi;mysql;odbc;opengl;oracle;pgsql;pyroot;pyroot2;pyroot3;pyroot_legacy;pythia6;pythia6_nolink;pythia8;qt5web;r;roofit;shadowpw;spectrum;sqlite;ssl;test_distrdf_pyspark;tmva;tmva-cpu;tmva-gpu;tmva-pymva;tmva-rmva;tmva-```

**Comment by bellenot** [comment]
@oshadura then why was it not failing before?

**Comment by oshadura** [comment]
Maybe my guess is wrong, checking other PR https://lcgapp-services.cern.ch/root-jenkins/job/root-pullrequests-build/121249/consoleFull I see that python3.8 is actually available on this node. Do you know if CMake was updated on this node as well?

**Comment by oshadura** [comment]
@phsft-bot build just on mac11.0/cxx17

**Comment by bellenot** [comment]
> Do you know if CMake was updated on this node as well?

No idea. @Axel-Naumann ?

**Comment by oshadura** [comment]
Yes cmake is updated, I will check: ```sftnight@macphsft23 ~ % cmake --version
cmake version 3.18.1
```

**Comment by oshadura** [comment]
@bellenot I think now PR is ready, before I made a small typo. The only problem is that all MacOS nodes need to have updated CMake (e.g. see macphsft17.dyndns.cern.ch)...

**Comment by bellenot** [comment]
@phsft-bot build

**Comment by bellenot** [comment]
@phsft-bot  build

**Review by bellenot — APPROVED** [review]

**Pull request merged**